java/ql/src/semmle/code/java/security/UnsafeDeserialization.qll

```diff
@@ -109,6 +176,119 @@ class SafeKryo extends DataFlow2::Configuration {
   }
 }
 
+class EnabledJacksonDefaultTyping extends DataFlow2::Configuration {
+  EnabledJacksonDefaultTyping() { this = "EnabledJacksonDefaultTyping" }
+
+  override predicate isSource(DataFlow::Node src) {
+    any(EnableJacksonDefaultTyping ma).getQualifier() = src.asExpr()
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof ObjectMapperReadSink }
+}
+
+class SafeObjectMapper extends DataFlow2::Configuration {
+  SafeObjectMapper() { this = "SafeObjectMapper" }
+
+  override predicate isSource(DataFlow::Node src) {
+    src instanceof SetPolymorphicTypeValidatorSource
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof ObjectMapperReadSink }
+
+  /**
+   * Holds if `fromNode` to `toNode` is a dataflow step
+   * that configures or creates an `ObjectMapper` via a builder.
+   */
+  override predicate isAdditionalFlowStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
+    exists(MethodAccess ma, Method m, Expr q | m = ma.getMethod() and q = ma.getQualifier() |
+      m.getDeclaringType() instanceof MapperBuilder and
+      m.getReturnType()
+          .(RefType)
+          .hasQualifiedName("com.fasterxml.jackson.databind.json",
+            ["JsonMapper$Builder", "JsonMapper"]) and
+      fromNode.asExpr() = q and
+      ma = toNode.asExpr()
+    )
+  }
+}
+
+class UnsafeType extends TaintTracking2::Configuration {
+  UnsafeType() { this = "UnsafeType" }
+
+  override predicate isSource(DataFlow::Node src) { src instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(MethodAccess ma, int i, Expr arg | i > 0 and ma.getArgument(i) = arg |
+      ma.getMethod() instanceof ObjectMapperReadMethod and
+      arg.getType() instanceof JacksonType and
+      arg = sink.asExpr()
+    )
+  }
+
+  /**
+   * Holds if `fromNode` to `toNode` is a dataflow step that looks like resolving a class.
+   */
+  override predicate isAdditionalTaintStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
+    exists(MethodAccess ma, RefType returnType | returnType = ma.getMethod().getReturnType() |
+      returnType instanceof JacksonType and
+      ma.getAnArgument() = fromNode.asExpr() and
+      ma = toNode.asExpr()
+    )
+  }
+}
+
+/**
+ * Holds if `fromNode` to `toNode` is a dataflow step that creates a Jackson parser.
+ */
+predicate createJacksonJsonParserStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
```

Comment:
Could you give an end-to-end example that these steps are designed to address?

Reply:
`ObjectMapper` has multiple `readValue()` and `readValues()` methods that take a `JsonParser`. In turn, `JsonParser` can take JSON from untrusted sources. This predicate defines steps that create a `JsonParser`. Here is an example:

```java
    private static <T> List<T> deserialize(String string, Class<T> clazz)
            throws IOException {

        ObjectMapper mapper = new ObjectMapper();
        mapper.enableDefaultTyping();
        JsonParser parser = new JsonFactory().createParser(string);
        return mapper.readValues(parser, clazz).readAll();
    }
```

Reply:
> Could the `createParser` be a sink, 

I don't think so. `JsonParser` parses JSON and returns primitive values, strings, JSON tokens, etc. It can return a type id or descriptor but as far as I know the parser doesn't create instances of these types.

> or could we still encounter sanitisers on the road to `readValues`?

I am aware of only two reliable sanitizers that are covered in the query. I was considering adding `ObjectMapper.activateDefaultTyping()` because it takes `PolymorphicTypeValidator`. But it turned out that in this case the validator is not always called. I reported this to the project, but they said that's expected. Please see https://github.com/FasterXML/jackson-databind/issues/2524 . Therefore, I didn't add this method as a sanitizer.